api/lorawan.proto

```diff
@@ -393,136 +401,160 @@ message MACCommand {
   }
   message LinkCheckAns {
+    option (gogoproto.populate) = false;
+
     // Indicates the link margin in dB of the received LinkCheckReq, relative to the demodulation floor.
-    uint32 margin = 1;
-    uint32 gateway_count = 2;
+    uint32 margin = 1 [(validate.rules).uint32.lte = 254];
+    uint32 gateway_count = 2 [(validate.rules).uint32.lte = 255];
   }
   message LinkADRReq {
-    DataRateIndex data_rate_index = 1;
-    uint32 tx_power_index = 2;
-    repeated bool channel_mask = 3;
+    option (gogoproto.populate) = false;
+
+    DataRateIndex data_rate_index = 1 [(validate.rules).enum.defined_only = true];
+    uint32 tx_power_index = 2 [(validate.rules).uint32.lte = 15];
+    repeated bool channel_mask = 3 [(validate.rules).repeated.max_items = 16];
     reserved 4; // For RFU 7th bit of Redundancy.
-    uint32 channel_mask_control = 5;
-    uint32 nb_trans = 6;
+    uint32 channel_mask_control = 5 [(validate.rules).uint32.lte = 7];
+    uint32 nb_trans = 6 [(validate.rules).uint32.lte = 15];
   }
   message LinkADRAns {
     bool channel_mask_ack = 1;
     bool data_rate_index_ack = 2;
     bool tx_power_index_ack = 3;
   }
   message DutyCycleReq {
-    AggregatedDutyCycle max_duty_cycle = 1;
+    AggregatedDutyCycle max_duty_cycle = 1 [(validate.rules).enum.defined_only = true];
   }
   message RxParamSetupReq {
-    DataRateIndex rx2_data_rate_index = 1;
-    uint32 rx1_data_rate_offset = 2;
-    uint64 rx2_frequency = 3; // Rx2 frequency (Hz).
+    option (gogoproto.populate) = false;
+
+    DataRateIndex rx2_data_rate_index = 1 [(validate.rules).enum.defined_only = true];
+    uint32 rx1_data_rate_offset = 2 [(validate.rules).uint32.lte = 7];
+    uint64 rx2_frequency = 3 [(validate.rules).uint64 = {gte: 100000, lte: 1677721600}]; // Rx2 frequency (Hz).
   }
   message RxParamSetupAns {
     bool rx2_data_rate_index_ack = 1;
     bool rx1_data_rate_offset_ack = 2;
     bool rx2_frequency_ack = 3;
   }
   message DevStatusAns {
+    option (gogoproto.populate) = false;
+
     // Device battery status.
     // 0 indicates that the device is connected to an external power source.
     // 1..254 indicates a battery level.
     // 255 indicates that the device was not able to measure the battery level.
-    uint32 battery = 1;
+    uint32 battery = 1 [(validate.rules).uint32.lte = 255];
     // SNR of the last downlink (dB; [-32, +31]).
-    int32 margin = 2;
+    int32 margin = 2 [(validate.rules).int32 = {gte: -32, lte: 31}];
   }
   message NewChannelReq {
-    uint32 channel_index = 1;
-    uint64 frequency = 2; // Channel frequency (Hz).
-    DataRateIndex min_data_rate_index = 3;
-    DataRateIndex max_data_rate_index = 4;
+    option (gogoproto.populate) = false;
+
+    uint32 channel_index = 1 [(validate.rules).uint32.lte = 255];
+    uint64 frequency = 2 [(validate.rules).uint64 = {gte: 100000, lte: 1677721600}]; // Channel frequency (Hz).
```

Comment:
Idem, please check other occurrences

Reply:
This is LoRaWAN spec - we can't send anything beyond this range using this MAC command
![2019-03-14-16:47:36-screenshot](https://user-images.githubusercontent.com/12877905/54371030-0297ca00-4679-11e9-9999-9219739f7b7e.png)